### PR TITLE
Genericinterface pendingtime diff

### PR DESCRIPTION
**Pull request by fredericve**
With this change you can perform a request towards GenericInterface with the PendingTime->{Diff} attribute set to a numeric value.

This allows for easier pending time setting for pending states. Kernel/System/Ticket.pm already supports this. I'm just adding it for GenericInterface where it always required the <year><month><day><hour><minute> attributes before.

### CODE REVIEW

**Comment by carlosfrodriguez** [comment]
Hi @fredericve,

Thank you for your proposal, this is in deed interesting.

Could it be possible for you to also update TicketCreate and TicketUpdate operations POD, also the WSDL has to be updated.

((enjoy))
cr

File: Kernel/GenericInterface/Operation/Ticket/Common.pm [hunk 1]
```diff
@@ -727,6 +733,13 @@ sub ValidatePendingTime {
     return if !$Param{PendingTime};
     return if !IsHashRefWithData( $Param{PendingTime} );
 
+    # if only the Diff attribute is present, check if it's a valid number and return.
+    # Nothing else needs to be checked in that case.
+    if ( keys %{ $Param{PendingTime} } == 1 && exists( $Param{PendingTime}->{Diff} ) ) {
```
Comment (carlosfrodriguez) [inline Kernel/GenericInterface/Operation/Ticket/Common.pm]: Why use exists and not defined?

Provably you can use defined and you can remove the ()

:-)

((enjoy))
cr

File: Kernel/GenericInterface/Operation/Ticket/Common.pm [hunk 2]
```diff
@@ -727,6 +733,13 @@ sub ValidatePendingTime {
...
+        return if $Param{PendingTime}->{Diff} !~ /^[0-9]+$/;
```
Comment (carlosfrodriguez) [inline Kernel/GenericInterface/Operation/Ticket/Common.pm]: Might be better:
```perl
 return if $Param{PendingTime}->{Diff} !~ m{\A \d+ \z}msx;
```

((enjoy))
cr

File: Kernel/GenericInterface/Operation/Ticket/Common.pm (hunk 1 above)
Comment (fredericve) [inline Kernel/GenericInterface/Operation/Ticket/Common.pm]: Hello @carlosfrodriguez,

Thanks for the feedback!

From the perl defined() documentation:
When used on a hash element, it tells you whether the value is defined, not whether the key exists in the hash. Use exists for the latter purpose.

Hence why I used exists.

**Comment by fredericve** [comment]
Hello @carlosfrodriguez,

I updated the files you suggested as well.

Tested all the changes in our UAT environment.

Any other suggestions are welcomed!

File: Kernel/GenericInterface/Operation/Ticket/Common.pm [hunk 3]
```diff
@@ -727,6 +733,13 @@ sub ValidatePendingTime {
     return if !$Param{PendingTime};
     return if !IsHashRefWithData( $Param{PendingTime} );
 
+    # if only the Diff attribute is present, check if it's a valid number and return.
+    # Nothing else needs to be checked in that case.
+    if ( keys %{ $Param{PendingTime} } == 1 && exists $Param{PendingTime}->{Diff} ) {
```
Comment (carlosfrodriguez) [inline Kernel/GenericInterface/Operation/Ticket/Common.pm]: Hi @fredericve,

The problem with exists is that it return success even if you have Diff => undef, as the key exists in the hash, no matter if the value is undefined, then that is why I believe that what you really really what is to make sure that you have a value in the key even if it is 0, then the best option is to use defined.

Does this makes sense?

((enjoy))
cr

Comment (fredericve) [inline Kernel/GenericInterface/Operation/Ticket/Common.pm]: Hello @carlosfrodriguez 

Agreed! I have updated the pull request.

Cheers,

Frederic

**Comment by carlosfrodriguez** [comment]
Hi @fredericve,

I've merged your feature in https://github.com/OTRS/otrs/commit/273d6f3bf735095b6148373afd722821ffd40a67, I just add an extra check to forbid the use of Diff together with other keys and I add a couple of unit tests.

Thanks again for your contribution!

((enjoy))
cr

**Pull request closed without merging**